pkg/cli/clusteroperator/pernodepod/runtime.go

```diff
@@ -0,0 +1,242 @@
+			if err != nil {
+				fmt.Fprintf(r.ErrOut, "failed to cleanup namespace: %v", err)
+			}
+		}()
+
+		if prePodHookFn != nil {
+			cleanup, err := prePodHookFn(ctx, actualNamespace.Name)
+			if err != nil {
+				return fmt.Errorf("pre-pod hook failed: %w", err)
+			}
+			if cleanup != nil {
+				defer cleanup(ctx)
+			}
+		}
+	}
+
+	workCh := make(chan *corev1.Node, numberOfNodesInParallel)
+	// producer
+	go func(ctx context.Context, nodes []*corev1.Node) {
+		defer close(workCh)
+		for i := range nodes {
+			select {
+			case workCh <- nodes[i]:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}(ctx, interestingNodes)
+
+	// consumer
+	wg := sync.WaitGroup{}
+	errCh := make(chan error, len(interestingNodes))
+	for i := 0; i < numberOfNodesInParallel; i++ {
+		wg.Add(1)
+		go func(ctx context.Context) {
+			defer wg.Done()
+			for {
+				select {
+				case node, stillReady := <-workCh:
+					if !stillReady {
+						return
+					}
+					if restartErr := r.RestartNode(ctx, createPodFn, nsName, node); restartErr != nil {
+						errCh <- restartErr
+					}
+				case <-ctx.Done():
+					return
+				}
+			}
+		}(ctx)
+	}
+	wg.Wait()
+	close(errCh)
+	errs := []error{}
+	for err := range errCh {
+		errs = append(errs, err)
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (r *PerNodePodRuntime) RestartNode(ctx context.Context, createPodFn CreatePodFunc, namespaceName string, node *corev1.Node) error {
+	if r.DryRun {
+		fmt.Fprintf(r.Out, "node/%v kubelet restarted\n", node.Name)
+		return nil
+	}
+
+	// it should only take X minutes per node
+	timeLimitedCtx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	defer cancel()
+
+	restartObj, err := createPodFn(ctx, namespaceName, node.Name, r.ImagePullSpec)
+	if err != nil {
+		return fmt.Errorf("unable to create pod manifest: %w", err)
+	}
+	createdPod, err := r.KubeClient.CoreV1().Pods(namespaceName).Create(timeLimitedCtx, restartObj, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("node/%v failed failed creating pod in --namespace=%v: %w", node.Name, namespaceName, err)
+	}
+	//fmt.Fprintf(o.Out, "Restarting node/%v using --namespace=%v pod/%v\n", node.Name, namespaceName, createdPod.Name)
+
+	lastWatchEvent, watchErr := watchtools.UntilWithSync(timeLimitedCtx,
+		cache.NewListWatchFromClient(
+			r.KubeClient.CoreV1().RESTClient(), "pods", namespaceName, fields.OneTermEqualSelector("metadata.name", createdPod.Name)),
+		&corev1.Pod{},
+		nil,
+		conditions.PodDone,
+	)
+	if watchErr != nil {
+		// TODO inspect and report the pod state.
+		retErr := fmt.Errorf("node/%v failed waiting for restart using --namespace=%v pod/%v: %w", node.Name, namespaceName, createdPod.Name, watchErr)
+		fmt.Fprintln(r.ErrOut, retErr.Error())
+		return retErr
+	}
+	finalPodState := lastWatchEvent.Object.(*corev1.Pod)
+
+	switch {
+	case finalPodState.Status.Phase == corev1.PodSucceeded:
```

Comment:
The semantics here get a bit weird, because this pod is rebooting the node...so the pod can't enter "completed" status really, unless we do the reboot *asynchronously* of the pod exiting successfully, which is what I was arguing with `systemd-run`.

In the general case this problem of "reporting reboot success" is so messy that in other places what we've done is just to fire off a "reboot" command, and then poll the node's `/proc/sys/kernel/boot-id`.  In kube that's actually part of the node status object, see `bootID` in  https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#nodesysteminfo-v1-core

Reply:
> The semantics here get a bit weird, because this pod is rebooting the node...so the pod can't enter "completed"

The output suggests that we do indeed enter a succeeded state. 

```yaml
    state:
      terminated:
        containerID: cri-o://eed849055daccd50d2e730a8f5b8e694b0751c84c8597151ce1460f5620da49d
        exitCode: 0
        finishedAt: "2023-06-13T11:12:39Z"
        reason: Completed
        startedAt: "2023-06-13T11:12:33Z"
  hostIP: 10.0.128.4
  phase: Succeeded
 ```

Reply:
OK.  Just saying I've seen historical issues in this area.

Reply:
> OK. Just saying I've seen historical issues in this area.

I know we've had issues rebooting nodes, I think restarting kubelets will be ok for this  usage.